pkg/core/klinedriver/tick_klinedriver.go

```diff
@@ -0,0 +1,180 @@
+package klinedriver
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/c9s/bbgo/pkg/types"
+	log "github.com/sirupsen/logrus"
+)
+
+type TickKLineDriver struct {
+	symbol        string
+	tickInterval  types.Interval
+	intervals     map[types.Interval]struct{}
+	intervalSlice []types.Interval
+
+	tradeTimeLocation *time.Location
+	kLineEmitter      KLineEmitter
+
+	mu           sync.Mutex
+	builder      *KLineBuilder
+	running      bool
+	tradesBuffer []*types.Trade
+}
+
+type KLineEmitter interface {
+	EmitKLine(types.KLine)
+	EmitKLineClosed(types.KLine)
+}
+
+func NewTickKLineDriver(symbol string, tickInterval types.Interval) *TickKLineDriver {
+	return &TickKLineDriver{
+		symbol:        symbol,
+		tickInterval:  tickInterval,
+		intervals:     make(map[types.Interval]struct{}),
+		intervalSlice: []types.Interval{},
+		builder:       NewKLineBuilder(symbol),
+		running:       false,
+		tradesBuffer:  []*types.Trade{},
+	}
+}
+
+func (t *TickKLineDriver) SetKLineEmitter(emitter KLineEmitter) {
+	t.kLineEmitter = emitter
+}
+
+// SetRunning sets the running state of the driver.
+// As for backtesting, use this method to set the start time of the driver.
+func (t *TickKLineDriver) SetRunning(startTime time.Time) {
+	// add intervals
+	for interval := range t.intervals {
+		// initialize the kline with the truncated start time
+		startTimeTruncated := startTime.Truncate(interval.Duration())
+		t.builder.AddInterval(interval, types.Time(startTimeTruncated))
+	}
+	t.running = true
+}
+
+func (t *TickKLineDriver) Subscribe(interval types.Interval) error {
+	if interval.Duration() < t.tickInterval.Duration() {
+		return fmt.Errorf("interval %s can't be smaller than tick interval %s", interval, t.tickInterval)
+	}
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	if _, ok := t.intervals[interval]; ok {
+		return nil
+	}
+	t.intervals[interval] = struct{}{}
+	t.intervalSlice = append(t.intervalSlice, interval)
+	return nil
+}
+
+// AddTrade adds a trade to the kline driver.
+func (t *TickKLineDriver) AddTrade(trade types.Trade) {
```

Comment:
I saw you use `AddTrade(trade *types.Trade)` in the builder, you can make them consistent.  I think using `trade types.Trade` is fine for both methods.

Reply:
ok